add OPpCONST_LINE flag to aid deparsing

iabyn opened the pull request:
Code like

    my $line = __LINE__;

was being deparsed as

    my $line = '42';

because __LINE__ is converted by the lexer into an ordinary OP_CONST whose SV is just a plain string holding the line number.

This commit adds a flag, OPpCONST_LINE, to OP_CONST, which is set to indicate that this OP_CONST was created from a __LINE__ construct.

Deparse.pm then uses this new flag to print out the literal string "__LINE__" rather than the constant's actual value.

This fixes one failing test file under
    cd t; ./TEST -deparse
and removes one (but not all) sources of failure from three others.

tonycoz reviewed (approved):
The change itself looks file.

Does `__FILE__` need a similar change?

iabyn committed:
add OPpCONST_TOKEN_* flags to aid deparsing

Code like

    my $line = __LINE__;
    my $file = __FILE__;
    my $pkg  = __PACKAGE__;

was being deparsed as

    my $line = '42';
    my $file = 'foo.pl';
    my $pkg  = 'Foo::Bar';

because __LINE__ etc are converted by the lexer into an ordinary
OP_CONST, whose SV is just a plain string holding the line number or
whatever.

This commit uses two spare bits in op_private on OP_CONST ops to create
a 4-valued bit field, indicating whether the op is a normal constant or
one of those three special constant tokens. This field is then used
by Deparse to output a literal "__LINE__" or whatever rather than the
value of the constant.

This fixes one failing test file under
    cd t; ./TEST -deparse
and removes one (but not all) sources of failure from three others.

Note that __SUB__ is already special-cased to deparse literally
(except when used outside a sub where it deparses as 'undef'), while
the special tokens __DATA__, __END__ and __CLASS__ don't generate
OP_CONSTs.

tonycoz commented on lib/B/Deparse.t:
```diff
@@ -3391,3 +3391,6 @@ $x = (return 'ok');
 $x //= (return 'ok');
 $x = exit 42;
 $x //= exit 42;
+####
+# preserve __LINE__ etc
+my $x = __LINE__;
```
This could use tests for `__FILE__` and `__PACKAGE__` now that you're adding support for them.

iabyn commented:
Yeah, I meant to update the test but forgot. Done now, and rebased and repushed.
